libp2p/transports/tortransport.nim

```diff
@@ -0,0 +1,141 @@
+
+proc connHandler*(self: TorTransport,
+                  client: StreamTransport,
+                  dir: Direction): Future[Connection] {.async.} =
+  var observedAddr: MultiAddress = MultiAddress()
+  try:
+    observedAddr = MultiAddress.init("/ip4/0.0.0.0").tryGet()
+  except CatchableError as exc:
+    trace "Failed to create observedAddr", exc = exc.msg
+    if not(isNil(client) and client.closed):
+      await client.closeWait()
+    raise exc
+
+  trace "Handling tcp connection", address = $observedAddr,
+                                   dir = $dir,
+                                   clients = self.tcpTransport.clients[Direction.In].len +
+                                   self.tcpTransport.clients[Direction.Out].len
+
+  let conn = Connection(
+    ChronosStream.init(
+      client = client,
+      dir = dir,
+      observedAddr = observedAddr
+    ))
+
+  proc onClose() {.async.} =
+    try:
+      let futs = @[client.join(), conn.join()]
+      await futs[0] or futs[1]
+      for f in futs:
+        if not f.finished: await f.cancelAndWait() # cancel outstanding join()
+
+      trace "Cleaning up client", addrs = $client.remoteAddress,
+                                  conn
+
+      self.tcpTransport.clients[dir].keepItIf( it != client )
+      await allFuturesThrowing(
+        conn.close(), client.closeWait())
+
+      trace "Cleaned up client", addrs = $client.remoteAddress,
+                                 conn
+
+    except CatchableError as exc:
+      let useExc {.used.} = exc
+      debug "Error cleaning up client", errMsg = exc.msg, conn
+
+  self.tcpTransport.clients[dir].add(client)
+  asyncSpawn onClose()
+
+  return conn
+
+method dial*(
+  self: TorTransport,
+  hostname: string,
+  address: MultiAddress): Future[Connection] {.async, gcsafe.} =
+  ## dial a peer
+  ##
+
+  trace "Dialing remote peer", address = $address
+  let transp = await connect(self.transportAddress)
+  try:
+    var bytesWritten = await transp.write(@[05'u8, 01, 00])
+    var resp = await transp.read(2)
+
+    let addressArray = ($address).split('/')
+    let addressStr = addressArray[2].split(':')[0] & ".onion"
+
+    let port = string.fromBytes(address.data.buffer[37..38])
+
+    bytesWritten = await transp.write("\x05\x01\x00\x03" & addressStr.len.char & addressStr & port)
+    resp = await transp.read(10)
+    echo resp
+
+    return await self.connHandler(transp, Direction.Out)
+  except CatchableError as err:
+    await transp.closeWait()
+    raise err
+
+method start*(
+  self: TorTransport,
+  addrs: seq[MultiAddress]) {.async.} =
+  ## listen on the transport
+  ##
+
+  await self.tcpTransport.start(addrs)
+
+method accept*(self: TorTransport): Future[Connection] {.async, gcsafe.} =
+  ## accept a new TCP connection
+  ##
+  return await self.tcpTransport.accept()
+
+method stop*(self: TorTransport) {.async, gcsafe.} =
+  ## stop the transport
+  ##
+  await self.tcpTransport.stop()
+
+
```

Comment:
You seem to be missing the `handles` method which helps the switch filter which transport to use for an address

Reply:
Done